generator/src/main/java/com/scottlogic/deg/generator/decisiontree/ConstraintNodeBuilder.java

```diff
@@ -0,0 +1,107 @@
+package com.scottlogic.deg.generator.decisiontree;
+
+import com.scottlogic.deg.common.profile.constraints.atomic.AtomicConstraint;
+import com.scottlogic.deg.common.util.FlatMappingSpliterator;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static java.util.stream.Stream.concat;
+
+public class ConstraintNodeBuilder {
+    private final Collection<AtomicConstraint> atomicConstraints;
+    private final Collection<DecisionNode> decisions;
+    private final Set<NodeMarking> nodeMarkings;
+
+    public ConstraintNodeBuilder(Collection<AtomicConstraint> atomicConstraints, Collection<DecisionNode> decisions, Set<NodeMarking> nodeMarkings) {
+        this.atomicConstraints = atomicConstraints;
+        this.decisions = decisions;
+        this.nodeMarkings = nodeMarkings;
+    }
+
+    public ConstraintNodeBuilder() {
+        this(Collections.emptyList(), Collections.emptyList(), Collections.emptySet());
+    }
+
+    public ConstraintNodeBuilder setAtomicConstraints(Collection<AtomicConstraint> newAtomicConstraints) {
+        return new ConstraintNodeBuilder(newAtomicConstraints, decisions, nodeMarkings);
+    }
+
+    public ConstraintNodeBuilder removeAtomicConstraint(AtomicConstraint atomicConstraint) {
+        return setAtomicConstraints(atomicConstraints
+            .stream()
+            .filter(constraint -> !constraint.equals(atomicConstraint))
+            .collect(Collectors.toList())
+        );
+    }
+
+    public ConstraintNodeBuilder addAtomicConstraints(Collection<AtomicConstraint> constraints) {
+        return setAtomicConstraints(
+            concat(
+                atomicConstraints.stream(),
+                constraints.stream()
+            ).collect(Collectors.toList())
+        );
```

Comment:
It might be clearer to write this method like this:
```
    public ConstraintNodeBuilder addAtomicConstraints(Collection<AtomicConstraint> atomicConstraints) {
        return setAtomicConstraints(
            concat(
                this.atomicConstraints.stream(),
                atomicConstraints.stream()
            ).collect(Collectors.toList())
        );
```

There are similar methods below.